### ISSUE TITLE
[BUG] Shoot kubeconfig reloading seems to not work in all cases

### ISSUE THREAD

**Issue by rfranzke** [gardener/dependency-watchdog#36]
**Describe the bug:**
After https://github.com/gardener/gardener/pull/5011 we rarely/sporadically see that the `dependency-watchdog-probe` does sometimes not reload the correct kubeconfig for communication with the shoot's API server.

Concretely, in my observations the `shoot-access-dependency-watchdog-external-probe` was affected (while the `shoot-access-dependency-watchdog-internal-probe` secret was working fine). With above PR, the kubeconfigs stored in these secrets are no longer static but change every ~`12h`.

As a consequence, the KCM was scaled down for the respective shoot. By the way, this can be improved as well: Instead of scaling down KCM just because the external probe errored, we should be more precise which kind of error occurred. In above case, we could actually communicate with KAPI via the external kubeconfig, we were just not authenticated, so scaling down is actually undesired.

**Expected behaviour:**

**How To Reproduce (as minimally and precisely as possible):**
I don't know, unfortunately.

**Logs:**
I don't have concrete logs (yet). However, the error was something like "the server asked to provide credentials" for the `external` probe while the `internal` probe was succeeding.

**Screenshots (if applicable):**
I don't have screenshots (yet).

**Anything else we need to know?**:
I had a rough look into the code and something around https://github.com/gardener/dependency-watchdog/blob/96f73d5dc311958f5cb6e914fbda10b8bbc095c9/pkg/scaler/prober.go#L78-L130 might be related.

**Comment by rfranzke** [comment]
Here are some concrete logs:

```
I1216 05:46:18.208883       1 scaler.go:72] Update event on cluster: shoot--it--tmg0b-o9o
I1216 05:46:25.150181       1 prober.go:360] shoot-kube-apiserver/shoot--it--tmg0b-o9o/internal: probe succeeded
I1216 05:46:25.150206       1 prober.go:375] shoot-kube-apiserver/shoot--it--tmg0b-o9o/internal: probe result: &scaler.probeResult{lastError:error(nil), resultRun:4}
I1216 05:46:25.151581       1 prober.go:363] shoot-kube-apiserver/shoot--it--tmg0b-o9o/external: probe failed with error: the server has asked for the client to provide credentials. Will retry...
I1216 05:46:25.151609       1 prober.go:375] shoot-kube-apiserver/shoot--it--tmg0b-o9o/external: probe result: &scaler.probeResult{lastError:(*errors.StatusError)(0xc0076e3c20), resultRun:4}
I1216 05:46:25.151636       1 prober.go:405] Scaling down dependents of shoot-kube-apiserver/shoot--it--tmg0b-o9o: apps/v1.Deployment/kube-controller-manager: replicas=0: in progress...
I1216 05:46:25.151655       1 prober.go:420] Scaling down dependents of shoot-kube-apiserver/shoot--it--tmg0b-o9o: apps/v1.Deployment/kube-controller-manager: skipped because desired=0 and current=0
```

Checking the respective kubeconfig yields to the fact it perfectly works:

```
$ kubectl --kubeconfig=<(kubectl -n shoot--it--tmg0b-o9o get secret shoot-access-dependency-watchdog-external-probe -o jsonpath={.data.kubeconfig} | base64 -d) get node
Error from server (Forbidden): nodes is forbidden: User "system:serviceaccount:kube-system:dependency-watchdog-external-probe" cannot list resource "nodes" in API group "" at the cluster scope
```

DWD has no RBAC permissions to list the nodes, however, it is authenticated against the `kube-apiserver`.

cc @ashwani2k @shreyas-s-rao

**Comment by ashwani2k** [comment]
Thanks @rfranzke  for providing detailed error log. 
I can concur with your observation. Currently the external probe [does not check for the kind of error ](https://github.com/gardener/dependency-watchdog/blob/96f73d5dc311958f5cb6e914fbda10b8bbc095c9/pkg/scaler/prober.go#L360)while deciding the [api-server being unhealthy. ](https://github.com/gardener/dependency-watchdog/blob/96f73d5dc311958f5cb6e914fbda10b8bbc095c9/pkg/scaler/prober.go#L279). 

This can be improved as a quick win to ensure KCM or any dependent component is not brought down for just any kind of error but only when kube-apiserver is unreachable. 

Currently the api-machinery liss [following statusReason for errors](https://github.com/kubernetes/apimachinery/blob/ea11419e6b79342dfdb688604cedc8c6ac84c5c3/pkg/api/errors/errors.go#L47)
```
	metav1.StatusReasonUnauthorized:        
	metav1.StatusReasonForbidden:          
	metav1.StatusReasonNotFound:              
	metav1.StatusReasonAlreadyExists:        
	metav1.StatusReasonConflict:              
	metav1.StatusReasonGone:                  
	metav1.StatusReasonInvalid:               
	metav1.StatusReasonServerTimeout: 
	metav1.StatusReasonTimeout:            
	metav1.StatusReasonTooManyRequests: 
	metav1.StatusReasonBadRequest:           
	metav1.StatusReasonMethodNotAllowed:
	metav1.StatusReasonNotAcceptable:       
	metav1.StatusReasonRequestEntityTooLarge: 
	metav1.StatusReasonUnsupportedMediaType: 
	metav1.StatusReasonInternalError:         
	metav1.StatusReasonExpired:               
	metav1.StatusReasonServiceUnavailable:   
```
of which I could isolate the following 2 can be checked to confirm if apiserver is unreachable -  
```
	metav1.StatusReasonServerTimeout:         
	metav1.StatusReasonServiceUnavailable:    
```

Would you agree that will be sufficient or should we also consider following as well 
```
	metav1.StatusReasonTimeout:               
	metav1.StatusReasonInternalError:       

```

cc @shreyas-s-rao

**Comment by rfranzke** [comment]
The "main" bug reported in this issue was that DWD did not refresh the kubeconfig for the shoot after it changed. Probably we should first focus on this one.
The fact that it does not take educated decisions based on the returned error when the communication to a shoot's kube-apiserver fails was just another random observation. Perhaps it should be put into its own issue and tackled after the first problem was fixed.

**Comment by ashwani2k** [comment]
Some more insights on the issue. 
We could see an occurrence for a cluster where the probe had failed around the time of secret rotation 
Checking the logs for the affected shoots below we can see that- 
1. First occurrence of the failure was to fetch the secret for external probe
2. Post that context was unregistered. 
3. A new probe was started with registering the context. However this time the internal probe failed with error `internal: probe failed with error: the server has asked for the client to provide credentials.`.
4. No unregistering of context was done this time. 
5. Post this 2 probes were started for the namespace.
6. Incidentally both registered the context and canceled the old context. 
7. Both probes succeeded the internal probe.
8. However only one succeeded the external probe the other didn't. 
9. After this each time the internal succeeded and external failed which was tried for 3 times after which the KCM was scaled down. 
10. Unfortunately the system didn't have logs beyond this time period for this shoot, so not sure of the actual event but normally whenever a secret is changed a new Probe for the namespace is initiated. 

I suspect that the context for the old probe was not cleanly cancelled. Need further investigation. 
This was followed with unregistering the context, but no cancel of old context was done. 
A new probe was started for namespace which failed with internal probe and as internal was unhealthy external was not tried
<details>
<summary>
Expand to see the logs 
</summary> 

```
  |   | 2022-01-17 19:55:37 | 4ERR | {"log":"Probe for namespace shoot--<project-name>--<shoot-name> returned error: secret \"shoot-access-dependency-watchdog-external-probe\" not found, \u0026{shoot-kube-apiserver 0xc000554b40 [0xc000554ac0]}","pid":"1","severity":"ERR","source":"scaler.go:291"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"Unregistering the context for the key: shoot--<project-name>--<shoot-name>/shoot-kube-apiserver","pid":"1","severity":"INFO","source":"multicontext.go:64"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"Starting the probe in the namespace shoot--<project-name>--<shoot-name>: \u0026{shoot-kube-apiserver 0xc000554b40 [0xc000554ac0]}","pid":"1","severity":"INFO","source":"scaler.go:263"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"Registering the context for the key: shoot--<project-name>--<shoot-name>/shoot-kube-apiserver","pid":"1","severity":"INFO","source":"multicontext.go:61"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe failed with error: the server has asked for the client to provide credentials. Will retry...","pid":"1","severity":"INFO","source":"prober.go:363"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe result: \u0026scaler.probeResult{lastError:(*errors.StatusError)(0xc015c0e8c0), resultRun:1}","pid":"1","severity":"INFO","source":"prober.go:375"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal is not healthy. Skipping the external probe.","pid":"1","severity":"INFO","source":"prober.go:331"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"Starting the probe in the namespace shoot--<project-name>--<shoot-name>: \u0026{shoot-kube-apiserver 0xc000554b40 [0xc000554ac0]}","pid":"1","severity":"INFO","source":"scaler.go:263"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"Starting the probe in the namespace shoot--<project-name>--<shoot-name>: \u0026{shoot-kube-apiserver 0xc000554b40 [0xc000554ac0]}","pid":"1","severity":"INFO","source":"scaler.go:263"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"Registering the context for the key: shoot--<project-name>--<shoot-name>/shoot-kube-apiserver","pid":"1","severity":"INFO","source":"multicontext.go:61"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"Cancelling older context for the key: shoot--<project-name>--<shoot-name>/shoot-kube-apiserver","pid":"1","severity":"INFO","source":"multicontext.go:69"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"Registering the context for the key: shoot--<project-name>--<shoot-name>/shoot-kube-apiserver","pid":"1","severity":"INFO","source":"multicontext.go:61"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"Cancelling older context for the key: shoot--<project-name>--<shoot-name>/shoot-kube-apiserver","pid":"1","severity":"INFO","source":"multicontext.go:69"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe succeeded","pid":"1","severity":"INFO","source":"prober.go:360"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe result: \u0026scaler.probeResult{lastError:error(nil), resultRun:1}","pid":"1","severity":"INFO","source":"prober.go:375"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe succeeded","pid":"1","severity":"INFO","source":"prober.go:360"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe result: \u0026scaler.probeResult{lastError:error(nil), resultRun:1}","pid":"1","severity":"INFO","source":"prober.go:375"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe succeeded","pid":"1","severity":"INFO","source":"prober.go:360"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe result: \u0026scaler.probeResult{lastError:error(nil), resultRun:1}","pid":"1","severity":"INFO","source":"prober.go:375"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe failed with error: the server has asked for the client to provide credentials. Will retry...","pid":"1","severity":"INFO","source":"prober.go:363"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe result: \u0026scaler.probeResult{lastError:(*errors.StatusError)(0xc018d0c3c0), resultRun:1}","pid":"1","severity":"INFO","source":"prober.go:375"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"Scaling up dependents of shoot-kube-apiserver/shoot--<project-name>--<shoot-name>: apps/v1.Deployment/kube-controller-manager: replicas=1: in progress...","pid":"1","severity":"INFO","source":"prober.go:405"}
  |   | 2022-01-17 19:55:37 | 4INFO | {"log":"Scaling up dependents of shoot-kube-apiserver/shoot--<project-name>--<shoot-name>: apps/v1.Deployment/kube-controller-manager: skipped because desired=1 and current=1","pid":"1","severity":"INFO","source":"prober.go:420"}
  |   | 2022-01-17 19:56:02 | 1INFO | {"log":"Update event on cluster: shoot--<project-name>--<shoot-name>","pid":"1","severity":"INFO","source":"scaler.go:72"}
  |   | 2022-01-17 19:56:12 | 1INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe succeeded","pid":"1","severity":"INFO","source":"prober.go:360"}
  |   | 2022-01-17 19:56:12 | 1INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe result: \u0026scaler.probeResult{lastError:error(nil), resultRun:2}","pid":"1","severity":"INFO","source":"prober.go:375"}
  |   | 2022-01-17 19:56:12 | 1INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe failed with error: the server has asked for the client to provide credentials. Will retry...","pid":"1","severity":"INFO","source":"prober.go:363"}
  |   | 2022-01-17 19:56:12 | 1INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe result: \u0026scaler.probeResult{lastError:(*errors.StatusError)(0xc0088b4be0), resultRun:2}","pid":"1","severity":"INFO","source":"prober.go:375"}
  |   | 2022-01-17 19:56:32 | 2INFO | {"log":"Update event on cluster: shoot--<project-name>--<shoot-name>","pid":"1","severity":"INFO","source":"scaler.go:72"}
  |   | 2022-01-17 19:56:47 | 3INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe succeeded","pid":"1","severity":"INFO","source":"prober.go:360"}
  |   | 2022-01-17 19:56:47 | 3INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe result: \u0026scaler.probeResult{lastError:error(nil), resultRun:3}","pid":"1","severity":"INFO","source":"prober.go:375"}
  |   | 2022-01-17 19:56:47 | 3INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe failed with error: the server has asked for the client to provide credentials. Will retry...","pid":"1","severity":"INFO","source":"prober.go:363"}
  |   | 2022-01-17 19:56:47 | 3INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe result: \u0026scaler.probeResult{lastError:(*errors.StatusError)(0xc00bb79360), resultRun:3}","pid":"1","severity":"INFO","source":"prober.go:375"}
  |   | 2022-01-17 19:56:47 | 3INFO | {"log":"Scaling down dependents of shoot-kube-apiserver/shoot--<project-name>--<shoot-name>: apps/v1.Deployment/kube-controller-manager: replicas=0: in progress...","pid":"1","severity":"INFO","source":"prober.go:405"}
  |   | 2022-01-17 19:56:47 | 3INFO | {"log":"Scaling down dependents of shoot-kube-apiserver/shoot--<project-name>--<shoot-name>: apps/v1.Deployment/kube-controller-manager: replicas=0: successful","pid":"1","severity":"INFO","source":"prober.go:476"}
  |   | 2022-01-17 19:57:02 | 2INFO | {"log":"Update event on cluster: shoot--<project-name>--<shoot-name>","pid":"1","severity":"INFO","source":"scaler.go:72"}
  |   | 2022-01-17 19:57:19 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe succeeded","pid":"1","severity":"INFO","source":"prober.go:360"}
  |   | 2022-01-17 19:57:19 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe result: \u0026scaler.probeResult{lastError:error(nil), resultRun:4}","pid":"1","severity":"INFO","source":"prober.go:375"}
  |   | 2022-01-17 19:57:19 | 4INFO | {"log":"shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe failed with error: the server has asked for the client to provide credentials. Will retry...","pid":"1","severity":"INFO","source":"prober.go:363"}
```

**Comment by ashwani2k** [comment]
On simulating the same behavior on a dev cluster we see that if we delete-create or delete the secret for external probe the flow will trigger unregistering and cancellation of old context before starting the new probe for the namespace. 
This is different from the behavior observed in the logs from Canary where it only unregistered after encountering secrets not found error and didn't cancel the old context before starting the new probe on the namespace. 

<details>
<summary> Logs when triggered deletion and creation of secret in quick succession </summary>

```
2022-01-19 13:06:31
2
{"log":"2022-01-19T07:36:31.209905059Z stderr F I0119 07:36:31.209728 1 scaler.go:72] Update event on cluster: shoot--<project-name>--<shoot-name>"}
2022-01-19 13:06:34
2
{"log":"2022-01-19T07:36:34.908245855Z stderr F E0119 07:36:34.908132 1 scaler.go:291] Probe for namespace shoot--<project-name>--<shoot-name> returned error: secret \"shoot-access-dependency-watchdog-external-probe\" not found, \u0026{shoot-kube-apiserver 0xc000258240 [0xc0002581c0]}"}
2022-01-19 13:06:34
3
{"log":"2022-01-19T07:36:34.908254276Z stderr F I0119 07:36:34.908148 1 multicontext.go:64] Unregistering the context for the key: shoot--<project-name>--<shoot-name>/shoot-kube-apiserver"}
2022-01-19 13:06:34
3
{"log":"2022-01-19T07:36:34.908260394Z stderr F I0119 07:36:34.908152 1 multicontext.go:69] Cancelling older context for the key: shoot--<project-name>--<shoot-name>/shoot-kube-apiserver"}
2022-01-19 13:06:39
3
{"log":"2022-01-19T07:36:39.458457816Z stderr F I0119 07:36:39.458354 1 scaler.go:263] Starting the probe in the namespace shoot--<project-name>--<shoot-name>: \u0026{shoot-kube-apiserver 0xc000258240 [0xc0002581c0]}"}
2022-01-19 13:06:39
3
{"log":"2022-01-19T07:36:39.458565084Z stderr F I0119 07:36:39.458481 1 multicontext.go:61] Registering the context for the key: shoot--<project-name>--<shoot-name>/shoot-kube-apiserver"}
2022-01-19 13:06:39
3
{"log":"2022-01-19T07:36:39.460367062Z stderr F I0119 07:36:39.460312 1 prober.go:360] shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe succeeded"}
2022-01-19 13:06:39
3
{"log":"2022-01-19T07:36:39.460377833Z stderr F I0119 07:36:39.460328 1 prober.go:375] shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe result: \u0026scaler.probeResult{lastError:error(nil), resultRun:1}"}Show context
2022-01-19 13:06:39
3
{"log":"2022-01-19T07:36:39.462179482Z stderr F I0119 07:36:39.462121 1 prober.go:360] shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe succeeded"}
2022-01-19 13:06:39
3
{"log":"2022-01-19T07:36:39.462192933Z stderr F I0119 07:36:39.462135 1 prober.go:375] shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe result: \u0026scaler.probeResult{lastError:error(nil), resultRun:1}"}
2022-01-19 13:06:39
3
{"log":"2022-01-19T07:36:39.462198265Z stderr F I0119 07:36:39.462150 1 prober.go:405] Scaling up dependents of shoot-kube-apiserver/shoot--<project-name>--<shoot-name>: apps/v1.Deployment/kube-controller-manager: replicas=1: in progress..."}
2022-01-19 13:06:39
3
{"log":"2022-01-19T07:36:39.462214952Z stderr F I0119 07:36:39.462161 1 prober.go:420] Scaling up dependents of shoot-kube-apiserver/shoot--<project-name>--<shoot-name>: apps/v1.Deployment/kube-controller-manager: skipped because desired=1 and current=1"}
2022-01-19 13:07:01
1
{"log":"2022-01-19T07:37:01.210184663Z stderr F I0119 07:37:01.210073 1 scaler.go:72] Update event on cluster: shoot--<project-name>--<shoot-name>"}
``` 
</details>

<details>
<summary> Logs when just deleting the external probe secret and not creating it fresh </summary> 

```
022-01-19 13:11:53
2
{"log":"2022-01-19T07:40:02.001781752Z stderr F I0119 07:40:02.001209 1 scaler.go:72] Update event on cluster: shoot--<project-name>--<shoot-name>"}
2022-01-19 13:11:53
3
{"log":"2022-01-19T07:40:25.014374153Z stderr F I0119 07:40:25.014262 1 prober.go:360] shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe succeeded"}
2022-01-19 13:11:53
3
{"log":"2022-01-19T07:40:25.014404813Z stderr F I0119 07:40:25.014279 1 prober.go:375] shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/internal: probe result: \u0026scaler.probeResult{lastError:error(nil), resultRun:4}"}
2022-01-19 13:11:53
3
{"log":"2022-01-19T07:40:25.016008017Z stderr F I0119 07:40:25.015936 1 prober.go:360] shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe succeeded"}
2022-01-19 13:11:53
3
{"log":"2022-01-19T07:40:25.016023547Z stderr F I0119 07:40:25.015948 1 prober.go:375] shoot-kube-apiserver/shoot--<project-name>--<shoot-name>/external: probe result: \u0026scaler.probeResult{lastError:error(nil), resultRun:4}"}
2022-01-19 13:11:53
3
{"log":"2022-01-19T07:40:25.016028486Z stderr F I0119 07:40:25.015964 1 prober.go:405] Scaling up dependents of shoot-kube-apiserver/shoot--<project-name>--<shoot-name>: apps/v1.Deployment/kube-controller-manager: replicas=1: in progress..."}
2022-01-19 13:11:53
3
{"log":"2022-01-19T07:40:25.016032574Z stderr F I0119 07:40:25.015976 1 prober.go:420] Scaling up dependents of shoot-kube-apiserver/shoot--<project-name>--<shoot-name>: apps/v1.Deployment/kube-controller-manager: skipped because desired=1 and current=1"}
2022-01-19 13:11:53
4
{"log":"2022-01-19T07:40:25.037481849Z stderr F E0119 07:40:25.037402 1 scaler.go:291] Probe for namespace shoot--<project-name>--<shoot-name> returned error: secret \"shoot-access-dependency-watchdog-external-probe\" not found, \u0026{shoot-kube-apiserver 0xc000258240 [0xc0002581c0]}"}
2022-01-19 13:11:53
4
{"log":"2022-01-19T07:40:25.037488789Z stderr F I0119 07:40:25.037416 1 multicontext.go:64] Unregistering the context for the key: shoot--<project-name>--<shoot-name>/shoot-kube-apiserver"}
2022-01-19 13:11:53
4
{"log":"2022-01-19T07:40:25.037496397Z stderr F I0119 07:40:25.037420 1 multicontext.go:69] Cancelling older context for the key: shoot--<project-name>--<shoot-name>/shoot-kube-apiserver"}
2022-01-19 13:11:53
4
{"log":"2022-01-19T07:40:32.001881962Z stderr F I0119 07:40:32.001726 1 scaler.go:72] Update event on cluster: shoot--<project-name>--<shoot-name>"}
2022-01-19 13:11:53
1
{"log":"2022-01-19T07:41:02.00231178Z stderr F I0119 07:41:02.002253 1 scaler.go:72] Update event on cluster: shoot--<project-name>--<shoot-name>"}
2022-01-19 13:11:53
2
{"log":"2022-01-19T07:41:32.001929758Z stderr F I0119 07:41:32.001869 1 scaler.go:72] Update event on cluster: shoot--<project-name>--<shoot-name>"}Show context
```
</details>

**Comment by ashwani2k** [comment]
Insights from further investigation. 
We noticed from the logs captured in [36#issue comment](https://github.com/gardener/dependency-watchdog/issues/36#issuecomment-1016081090). 
Before 2 concurrent probes were started the external probe had failed with `secret \"shoot-access-dependency-watchdog-external-probe\" not found,` error. 
We noticed while testing locally  that when the external probe key is not found then it deletes the probe.  
As per the logs first the internal probe fails in the regular 30sec reconciliation flow. So it wasn't updated yet. 
But then it was updated and external probe wasn't updated. 
So we see internal succeed and external fails and the context is cancelled for the probe which was working with rotated keys and continued for the one with old token for external probe. 

So we see a possibility that as a result of rotation both the internal and external probe secrets were picked up at the same time and 2 probes were started.  
|Scenario|Internal Probe Secret| External Probe Secret| Behaviour|
|----------|---------------------|----------------------|-----------|
|1| - | Deleted| External probe fails and deletes the current probe|
|2| Secret Changed | Secret Added |  2 probes are started |
|3| Secret Changed | Secret is getting  Changed | 2 probes are started, however the secret of external probe ctx is still working with the old secrets| 

As the current logs are not sufficient to even identify what operations had initiated the flow and in what sequence the probes are created and deleted we have introduced more logs with the PR[#39](https://github.com/gardener/dependency-watchdog/pull/39) to get a better understanding of the behaviour.

**Comment by rfranzke** [comment]
It seems the issue got fixed with https://github.com/gardener/dependency-watchdog/pull/41, so let's
/close
